File: evaluation/S0/java/Hard/1876-contain-virus.java
// Source: https://leetcode.com/problems/contain-virus/   |   Difficulty: Hard
//
// Problem Description:
// A virus is spreading rapidly, and your task is to quarantine the infected area by installing walls.
//
// The world is modeled as an m x n binary grid isInfected, where isInfected[i][j] == 0 represents uninfected cells, and isInfected[i][j] == 1 represents cells contaminated with the virus. A wall (and only one wall) can be installed between any two 4-directionally adjacent cells, on the shared boundary.
//
// Every night, the virus spreads to all neighboring cells in all four directions unless blocked by a wall. Resources are limited. Each day, you can install walls around only one region (i.e., the affected area (continuous block of infected cells) that threatens the most uninfected cells the following night). There will never be a tie.
//
// Return the number of walls used to quarantine all the infected regions. If the world will become fully infected, return the number of walls used.
//
// Example:
// Input: isInfected = [[0,1,0,0,0,0,0,1],[0,1,0,0,0,0,0,1],[0,0,0,0,0,0,0,1],[0,0,0,0,0,0,0,0]]
// Output: 10
// Explanation: There are 2 contaminated regions.
// On the first day, add 5 walls to quarantine the viral region on the left. The board after the virus spreads is:
//
// On the second day, add 5 walls to quarantine the viral region on the right. The virus is fully contained.
//
// Constraints:
// m == isInfected.length
// 	n == isInfected[i].length
// 	1 <= m, n <= 50
// 	isInfected[i][j] is either 0 or 1.
// 	There is always a contiguous viral region throughout the described process that will infect strictly more uncontaminated squares in the next round.
//

class Solution {
    public int containVirus(int[][] isInfected) {
        int m = isInfected.length, n = isInfected[0].length;
        int totalWalls = 0;

        while (true) {
            boolean[][] visited = new boolean[m][n];
            List<Set<int[]>> regions = new ArrayList<>();
            List<Set<int[]>> frontiers = new ArrayList<>();
            List<Integer> wallsNeeded = new ArrayList<>();

            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (isInfected[i][j] == 1 && !visited[i][j]) {
                        Set<int[]> region = new HashSet<>();
                        Set<int[]> frontier = new HashSet<>();
                        int[] walls = new int[1];
                        dfs(isInfected, visited, i, j, region, frontier, walls);
                        regions.add(region);
                        frontiers.add(frontier);
                        wallsNeeded.add(walls[0]);
                    }
                }
            }

            if (regions.isEmpty()) break;

            int maxFrontierIdx = 0;
            for (int i = 1; i < frontiers.size(); i++) {
                if (frontiers.get(i).size() > frontiers.get(maxFrontierIdx).size()) {
                    maxFrontierIdx = i;
                }
            }

            totalWalls += wallsNeeded.get(maxFrontierIdx);

            for (int i = 0; i < regions.size(); i++) {
                if (i == maxFrontierIdx) {
                    for (int[] cell : regions.get(i)) {
                        isInfected[cell[0]][cell[1]] = -1;
                    }
                } else {
                    for (int[] cell : frontiers.get(i)) {
                        isInfected[cell[0]][cell[1]] = 1;
                    }
                }
            }
        }

        return totalWalls;          
    }
}

File: evaluation/S0/javascript/Hard/1876-contain-virus.js
// Source: https://leetcode.com/problems/contain-virus/   |   Difficulty: Hard
//
// Problem Description:
// A virus is spreading rapidly, and your task is to quarantine the infected area by installing walls.
//
// The world is modeled as an m x n binary grid isInfected, where isInfected[i][j] == 0 represents uninfected cells, and isInfected[i][j] == 1 represents cells contaminated with the virus. A wall (and only one wall) can be installed between any two 4-directionally adjacent cells, on the shared boundary.
//
// Every night, the virus spreads to all neighboring cells in all four directions unless blocked by a wall. Resources are limited. Each day, you can install walls around only one region (i.e., the affected area (continuous block of infected cells) that threatens the most uninfected cells the following night). There will never be a tie.
//
// Return the number of walls used to quarantine all the infected regions. If the world will become fully infected, return the number of walls used.
//
// Example:
// Input: isInfected = [[0,1,0,0,0,0,0,1],[0,1,0,0,0,0,0,1],[0,0,0,0,0,0,0,1],[0,0,0,0,0,0,0,0]]
// Output: 10
// Explanation: There are 2 contaminated regions.
// On the first day, add 5 walls to quarantine the viral region on the left. The board after the virus spreads is:
//
// On the second day, add 5 walls to quarantine the viral region on the right. The virus is fully contained.
//
// Constraints:
// m == isInfected.length
// 	n == isInfected[i].length
// 	1 <= m, n <= 50
// 	isInfected[i][j] is either 0 or 1.
// 	There is always a contiguous viral region throughout the described process that will infect strictly more uncontaminated squares in the next round.
//

/**
 * @param {number[][]} isInfected
 * @return {number}
 */
var containVirus = function(isInfected) {
	var m = isInfected.length;
	var n = isInfected[0].length;
	var dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]];
	var totalWalls = 0;

	while (true) {
		var visited = Array.from({ length: m }, function() {
			return Array(n).fill(false);
		});
		var regions = [];
		var frontiers = [];
		var wallsNeeded = [];

		for (var r = 0; r < m; r++) {
			for (var c = 0; c < n; c++) {
				if (isInfected[r][c] !== 1 || visited[r][c]) {
					continue;
				}

				var region = [];
				var frontier = new Set();
				var walls = 0;
				var stack = [[r, c]];
				visited[r][c] = true;

				while (stack.length > 0) {
					var cell = stack.pop();
					var cr = cell[0];
					var cc = cell[1];
					region.push([cr, cc]);

					for (var d = 0; d < 4; d++) {
						var nr = cr + dirs[d][0];
						var nc = cc + dirs[d][1];
						if (nr < 0 || nr >= m || nc < 0 || nc >= n) {
							continue;
						}
						if (isInfected[nr][nc] === 0) {
							walls += 1;
							frontier.add(nr + "," + nc);
						} else if (isInfected[nr][nc] === 1 && !visited[nr][nc]) {
							visited[nr][nc] = true;
							stack.push([nr, nc]);
						}
					}
				}

				regions.push(region);
				frontiers.push(frontier);
				wallsNeeded.push(walls);
			}
		}

		if (regions.length === 0) {
			break;
		}

		var maxIndex = -1;
		var maxFrontierSize = 0;
		for (var i = 0; i < frontiers.length; i++) {
			if (frontiers[i].size > maxFrontierSize) {
				maxFrontierSize = frontiers[i].size;
				maxIndex = i;
			}
		}

		if (maxFrontierSize === 0) {
			break;
		}

		totalWalls += wallsNeeded[maxIndex];

		for (var qi = 0; qi < regions[maxIndex].length; qi++) {
			var qcell = regions[maxIndex][qi];
			isInfected[qcell[0]][qcell[1]] = 2;
		}

		for (var ri = 0; ri < regions.length; ri++) {
			if (ri === maxIndex) {
				continue;
			}
			frontiers[ri].forEach(function(key) {
				var parts = key.split(",");
				var fr = parseInt(parts[0], 10);
				var fc = parseInt(parts[1], 10);
				if (isInfected[fr][fc] === 0) {
					isInfected[fr][fc] = 1;
				}
			});
		}
	}

	return totalWalls;
};
